File: typescript/src/calisthenics-exercises/only-one-dot-per-line.ts
class Address {
  private readonly street: string;
  private readonly city: string;

  constructor(street: string, city: string) {
    this.street = street;
    this.city = city;
  }

  getCity(): string {
    return this.city
  }
}

class Customer {
  private readonly name: string;
  private readonly address: Address;

  constructor(name: string, address: Address) {
    this.name = name;
    this.address = address;
  }

  getAddress(): Address {
    return this.address
  }
}

class Order {
  private readonly customer: Customer;

  constructor(customer: Customer) {
    this.customer = customer;
  }

  getCustomer(): Customer {
    return this.customer
  }
}

// Ejemplo de uso:

const order = new Order(new Customer('John Doe', new Address('Elm Street', 'Madrid')))
// Por ejemplo, para seleccionar los pedidos por localidad de destino y asignar la ruta de transporte
const destination = order.getCustomer().getAddress().getCity()

File: csharp/src/calisthenics-exercises/OnlyOneDotPerLine.cs
namespace RefactoringAvanzado.CalisthenicExercises;

public class Address
{
    private readonly string street;
    private readonly string city;

    public Address(string street, string city)
    {
        this.street = street;
        this.city = city;
    }

    public string GetCity()
    {
        return city;
    }
}

public class Customer
{
    private readonly string name;
    private readonly Address address;

    public Customer(string name, Address address)
    {
        this.name = name;
        this.address = address;
    }

    public Address GetAddress()
    {
        return address;
    }
}

public class Order
{
    private readonly Customer customer;

    public Order(Customer customer)
    {
        this.customer = customer;
    }

    public Customer GetCustomer()
    {
        return customer;
    }
}

public static class OrderExample
{
    public static string GetOrderDestination()
    {
        var order = new Order(new Customer("John Doe", new Address("Elm Street", "Madrid")));
        var destination = order.GetCustomer().GetAddress().GetCity();
        return destination;
    }
}
